### ISSUE TITLE
How can I reconnect on any kind of error?

### ISSUE THREAD

**Issue by mihneadb** [Parsely/pykafka#336]
Hi,

I've been using pykafka with a producer that seems unreliable. I'm seeing multiple exceptions which I can't handle; to name a few:

* Error committing offsets for topic xxxxxx (errors: {<class 'pykafka.exceptions.NotCoordinatorForConsumer'>: [0, 4]})
* NoNodeError ((), {})

I've read around the issue tracker and the PRs, and I noticed that for the 1st one there might be a fix in place. I've upgraded to the latest version, hoping that the "raise worker exc in main thread" PR will solve my need, but it doesn't look like it.
Basically, all I need is that whenever an exception happens (any exception, really), catch it and trigger a reconnect. I have the code for reconnecting, of course, I'm just having trouble catching all these exceptions.

What's the proper way to do this?
Thanks!

**Comment by emmett9001** [comment]
Hi @mihneadb, and thanks for checking this in. First, make sure that you're testing with current master, since this is an area we're actively working on improving, and there have been a bunch of fixes going in recently. Are you able to post full tracebacks for these errors? Often, pykafka will log an error before automatically retrying, so it's possible that you're running into that. Also it would allow us to make sure there are no thread death scenarios that aren't properly raising the associated exception.

**Comment by mihneadb** [comment]
Hi @emmett9001,
It's a bit more complicated checking against master since this is a multi-node setup - tldr it's easier to use something from pip, and I don't have control over the producers/brokers.

Here are some tracebacks from Sentry:

```
  File "raven/base.py", line 303, in build_msg
    transformer=self.transform)
  File "raven/base.py", line 459, in capture
    **kwargs)
  File "raven/handlers/logging.py", line 159, in _emit
    extra=extra, date=date, **kwargs)
  File "core/log/handlers/tagged_sentry_handler.py", line 34, in _emit
    return super(TaggedSentryHandler, self)._emit(record, **kwargs)
  File "raven/handlers/logging.py", line 59, in emit
    return self._emit(record)
  File "logging/__init__.py", line 749, in handle
    self.emit(record)
  File "logging/__init__.py", line 1321, in callHandlers
    hdlr.handle(record)
  File "logging/__init__.py", line 1281, in handle
    self.callHandlers(record)
  File "logging/__init__.py", line 1271, in _log
    self.handle(record)
  File "logging/__init__.py", line 1178, in error
    self._log(ERROR, msg, args, **kwargs)
  File "pykafka/simpleconsumer.py", line 407, in commit_offsets
    for err, parts in iteritems(parts_by_error)})
  File "pykafka/simpleconsumer.py", line 377, in _auto_commit
    self.commit_offsets()
  File "pykafka/simpleconsumer.py", line 301, in autocommitter
    self._auto_commit()
  File "python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "python2.7/threading.py", line 783, in __bootstrap
    self.__bootstrap_inner()
```

```
NoNodeError: ((), {})
  File "kazoo/recipe/watchers.py", line 333, in _get_children
    result = self._func(children)
  File "pykafka/balancedconsumer.py", line 599, in _brokers_changed
    self._rebalance()
  File "pykafka/balancedconsumer.py", line 493, in _rebalance
    self._remove_partitions(current_zk_parts - new_partitions)
  File "pykafka/balancedconsumer.py", line 528, in _remove_partitions
    self._zookeeper.delete(self._path_from_partition(p))
  File "kazoo/client.py", line 1293, in delete
    return self.delete_async(path, version).get()
  File "kazoo/handlers/utils.py", line 78, in get
    raise self._exception
```

Thanks!

**Comment by mihneadb** [comment]
Update: at least for the 2nd exception (NoNodeError), it is a legit problem -- after that, watching the offset lag, it's only increasing. So I have to manually restart the consumer.

**Comment by mihneadb** [comment]
Asked upstream. So when NoNodeError popped up, here's what they saw:

> yeah we saw some errors producing to kafka around then - looks like it swapped the leader for a partition
```
o.a.k.c.e.NotLeaderForPartitionException: This server is not the leader for that topic-partition.
```

**Comment by emmett9001** [comment]
@mihneadb Are you able to provide a test case that reliably reproduces the `NoNodeError` traceback? Also, since this exception is not raised on a thread, it seems like you should be able to wrap your code in a `try...except NoNodeError` to verify that doing so actually fixes the issue. Are either of those a possibility?

**Comment by mihneadb** [comment]
@emmett9001 I don't really know the code well enough. I think you should at least be able to reproduce the bug by using `mock` and patching the respective method to throw `NoNodeError`. Regarding wrapping in `try except`, I've already done that, with `except Exception` even and it did not catch it. So I think this actually happens in a thread.
Maybe the logging is not full (as I mentioned, it is a traceback that I got from sentry).

**Comment by emmett9001** [comment]
@yungchin what do you think about wrapping [this call](https://github.com/Parsely/pykafka/blob/640b2b581fb1bceadb1160638598ee9ca54cb7dc/pykafka/balancedconsumer.py#L568) in a `try...except: pass`, possibly with a warning? I think it makes sense because if we're removing the partition from ZK anyway, we don't mind if it's already gone for some reason. I imagine this is an edge case bug caused by a short-lived mismatch between ZK and the internally held partitions.

**Comment by yungchin** [comment]
@emmett9001 sorry I totally missed this!

I'm not sure - if this happens on a leader change, as per @mihneadb's comment, we may actually be doing the wrong thing if we ignore the `NoNodeError`: say we hold partition 0 at leader 0, and so our zk node ends in `topicname/0-0`, but the new leader is now leader 1, we'd get a `NoNodeError` because `_path_from_partition` (after an update) returns `topicname/1-0` - and so then we'd hold the node `topicname/0-0` forever? 

Hmmm, maybe it's worth revisiting that idea to store node versions...

**Comment by emmett9001** [comment]
@mihneadb Are you able to try reproducing this bug on version 2.2.0? A lot has changed since November.

**Comment by mihnea-dobrescubalaur-hs** [comment]
@emmett9001 sure thing, we're actually in the process of upgrading to the latest pykafka because we were hit by that memory leak that you recently fixed. :)

**Comment by emmett9001** [comment]
Thanks @mihnea-dobrescubalaur-hs, let us know here if you're able to reproduce the bug in this ticket with the latest version.

**Comment by mihneadb** [comment]
You got it!

On Mon, Feb 8, 2016 at 9:33 PM, Emmett J. Butler <notifications@github.com>
wrote:

> Thanks @mihnea-dobrescubalaur-hs
> <https://github.com/mihnea-dobrescubalaur-hs>, let us know here if you're
> able to reproduce the bug in this ticket with the latest version.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Parsely/pykafka/issues/336#issuecomment-181474982>.
>



-- 
Mihnea Dobrescu-Balaur
